test_runner/src/main/kotlin/ftl/run/AndroidTestRunner.kt

```diff
@@ -17,66 +18,69 @@ import kotlinx.coroutines.coroutineScope
 
 object AndroidTestRunner {
 
-    suspend fun runTests(androidArgs: AndroidArgs): MatrixMap = coroutineScope {
+    suspend fun runTests(androidArgs: AndroidArgs): Pair<MatrixMap, List<List<String>>> = coroutineScope {
         val (stopwatch, runGcsPath) = GenericTestRunner.beforeRunTests(androidArgs)
 
         // GcAndroidMatrix => GcAndroidTestMatrix
         // GcAndroidTestMatrix.execute() 3x retry => matrix id (string)
         val androidDeviceList = GcAndroidDevice.build(androidArgs.devices)
 
-        val apks = resolveApks(androidArgs, runGcsPath)
         val jobs = arrayListOf<Deferred<TestMatrix>>()
         val runCount = androidArgs.repeatTests
-        val shardCount = androidArgs.testShardChunks.size
         val shardCounter = ShardCounter()
         val history = GcToolResults.createToolResultsHistory(androidArgs)
+        val apks = resolveApks(androidArgs, runGcsPath)
+        val allTestShardChunks: MutableList<List<String>> = mutableListOf()
 
-        println(GenericTestRunner.beforeRunMessage(androidArgs))
-        repeat(runCount) {
-            repeat(shardCount) { testShardsIndex ->
-                jobs += async {
-                    GcAndroidTestMatrix.build(
-                        appApkGcsPath = apks.first,
-                        testApkGcsPath = apks.second,
-                        runGcsPath = runGcsPath,
-                        androidDeviceList = androidDeviceList,
-                        testShardsIndex = testShardsIndex,
-                        args = androidArgs,
-                        shardCounter = shardCounter,
-                        toolResultsHistory = history
-                    ).executeWithRetry()
+        apks.forEach { apk ->
+            // ensure we only shard tests that are part of the test apk
+            val testShardChunks = AndroidTestShard.getTestShardChunks(androidArgs, apk.test)
+            allTestShardChunks += testShardChunks
+            repeat(runCount) {
+                testShardChunks.forEach { testTargets ->
+                    jobs += async {
```

Comment:
We should probably specify a `Dispatcher` to run things in parallel because is inheriting the context from `runBlocking`

Reply:
Fixed